### ISSUE TITLE
from_xml of some federated objects fails

### ISSUE THREAD

**Issue by Raven24** [diaspora/diaspora#3519]
I think this currently causes most of our resque failures, and might swallow quite a chunk of federated objects...
I will walk through this step by step, I chose a Participation object, but I also got the same problem with Likes (and maybe other objects, too):

I received a `Jobs::ReceiveUnencryptedSalmon` with this payload:

```xml
<?xml version='1.0' encoding='UTF-8'?>
<diaspora xmlns="https://joindiaspora.com/protocol" xmlns:me="http://salmon-protocol.org/ns/magic-env">
  <header>    
    <author_id>deadsuperhero@joindiaspora.com</author_id>
  </header>
  <me:env>
    <me:data type='application/xml'>ICAgICAgICAgIDxYTUw-CiAgICAgICAgICA8cG9zdD48cGFydGljaXBhdGlvbj4KICA8Z3VpZD4xMTUwMTZjYjRlZGIyZWVlPC9ndWlkPgogIDx0YXJnZXRfdHlwZT5Qb3N0PC90YXJnZXRfdHlwZT4KICA8cGFyZW50X2d1aWQ-MWU3Y2FiNWY2ZmRkMTAyZTwvcGFyZW50X2d1aWQ-CiAgPHBhcmVudF9hdXRob3Jfc2lnbmF0dXJlPm92QncwK3h2cnpEQ0phbjBtYXFJYlhSTU9ZdXprckFjQWJFNEdwVytDQ2YrZXF6aVRLQjhINmxPZzZ4TkFHdjdtdWsxYkt4TGI1WjFNekNWeHZVMno5OVZKemZuSWpaVFR0aFRZd2Z3andjTFIyQmVobHJiU1ovazBCRE9VSS9SMkZFTWVLWFpXNndQWVhRRkFYYVNjSHFMQkZSdDBzdE1Qbk4wYkxCaWpuTWlFUTNaMWdYTmFUOFRDMlp3ZittdUVoS3Z3akQ1T3ZEMThpQmVyL2Nycy9SVXNkU1JFQjhGTklxR0wxckV4Mnc0WDZRRU9taWw4WHlIZjhVWldCUGRxWW1wRzlsQWM0U2NmTW00aUc4L1gxb3dHRU94a1JjOE10UFBmSXJ5YUV2Um96aEZ5a2luWlFaZE1pL0ZYbzV1YmVRZ0tPSmljdFgyS0JmMFlDc0RxSENoWGl2THpPR01XUkxNUXo5RUlsZHMzNzdac0QxOUdaS0lpckU4V3BKdkZ5eS9hUldTVkJuT1NVYlVJVnZIRitHSDI0R1VHcjRuQjVyd2xrazdwQWVmTDJIRTYrL2NFYWZFb2ZkVWtnSkN0RThmc3BsUVFKQy83UlpmVlB4TVZkSStqckdkOEY0WTlFZDB2VXUxSVJEQUphbGlJejFOakFUNFlHeHE3MFV4bG1wNGE2c2RBT1JGZC92RDc2Qk41dkNDYWVKZEhRd2lML1Z1a3RYTVBSMURGS09KclVMN1NsNkVBZXVYWmhxa0xIdTlVcVlINk1FVHEyK3ZzRlhCdGxWTUoyc3N1bUF0aVVCQURvWHhheExic3hxQWJtU2JNbC9aYlV3Sm1mTnByajFHTldVM3ZsdWFISHZtd1lOS1FJdUpJWFFVTUZqdEVHSi9tZGoxcFlZPTwvcGFyZW50X2F1dGhvcl9zaWduYXR1cmU-CiAgPGF1dGhvcl9zaWduYXR1cmU-UE5DOWpXYTRSTE5jUll5TVpSTlJJYmNIRUN4UFVtd0FjbW9iL2ZmWmZmR3pDdDROTDdycU5UME5VOXJSWXJRNkN3VU9aNm5tcmJETzZIc0JWanR5ZnJ0SXFyWEhvZCs1Y1NZUmhkT1ZCV1NaKzQ2YlIwU1NDQUtUYnN0WlVCQTR6ZVh5MVdPbU1ad0NUQnd2SmU2YWY4NldJSXZET2Fod1dHMHVpSENyRW5VYU5meVpnU2k0bU1JU1BPT0VEcnBDODVLajhMNm0xQldESkZmT2YzUFc1MlFreXhsM1hidDhUOVVNazMwNjVCYWdMRHJ1NUxHQ1NGWkdSbTZ6NVBjV0hmbGtZeFVMN242emluaXR4RW9SOGpNdkI1eHBseENnSXBOc1IzQUd2VGNOQkFhZ3FjZnNZTXRCLytGSEhIUTRQSkRBZy9Xb2RadDBkQlJvdnI5TlplNHgvNnhRTDNFNy9nM1dlakJ5K0F5NUZzSWNuWWZ5ZjdmbjV0UHhuY2NqWDB6YlBjRjIrVU9CK0YydUV4OVIzRGFIaXpCSFRaYUVDUnowTkhqMzdOYWs3WDdMT2NHV2hpemhnRnpsMlEra29vUnVsRC96UzlLdnNBZ1FrOSsrMEYwR1hOd3YyQ3FRWkYrWXB5eFc0VUw2NVBaMUhIZDNOMDFPR2RlRUVoQ2NxWHlGM3FFTUxUWkIxZkJVTm93M3B1bC92T05VcGF3WDErTk1Sc05CRXNFVnFOUFVkb3EvS0N4NWdTRkx3K0g1aVJxcVRDZTh2Q09RejRhbDhBblI3VTl3RWRXNnZZTS9vSzZqSFhmeWJnV09KTHpIWVBiQTBadmJremNlRldpUFlFL1JhTWY3MjYreVVzcHI4MTdZVWlvR1cxaGF1ZkVVT2lTL3ZOQ2x3SUU9PC9hdXRob3Jfc2lnbmF0dXJlPgogIDxkaWFzcG9yYV9oYW5kbGU-Y29wYWZsaWdodDhAam9pbmRpYXNwb3JhLmNvbTwvZGlhc3BvcmFfaGFuZGxlPgo8L3BhcnRpY2lwYXRpb24-PC9wb3N0PgogICAgICAgICAgPC9YTUw-Cg==</me:data>
    <me:encoding>base64url</me:encoding>
    <me:alg>RSA-SHA256</me:alg>
    <me:sig>NZAWFQZ7mUzizIhgQTkj3oRsx6Z4P-steiAfryWJsI6JYkdTB-MZUKcICjnp69uV2Udn4-0qrKd-SEaxEZENgoVzv8lX0AEbesucifyv3lRywLPQ9mw0tANESx0imp9q0qDAhpKtZMw9DL_J_YyiQ58omzYx2-OHlYG6LFHVgtg6yYhn51_Pb5-SNk5nPNvjoZTocbZikaWyYyuJbOLPuuBQsjv3OA4d8W-kl-loSWpnAvzrjIxgzxxZk_zZ6Tm3djO9ERvAJD6lOk1D9tiugq26jAnaicw-Z2hepRVqQU7gvOlv9JjszjhC8xBUX6RsNvoqDVnbmZI7cWGRP3EqP0V-ArLzzglV_a8rsHVOHHmHMnrW420N6hiZX6HOmVp4dliDBzFT8tyWnAicrQHPJ3fQQ9E1Yizm2X_iYmTlSyyrPopXO11IcLPpI2FvL1k_IxEF0PAQqDPMEuHMYqE4v8LaQZMebDfHZlVZDWZbn8gd-pVz4YdlGg06vq3RT0oVDdfz8Ep0AmkTmJ8LO1pg09GFCb5de9DEOd0WXX315gVWMX51nHnzybQc5pBa6_JGWsKHYQEl2IjQEF1mZ30PDsAu2_z0_sbvDMgmeMs7iYaafk453SAswvSS_Ty0zV4EUeKFRlYDkxZjv9gbmd1hQ7k6j9OJhcEXigmVqZ33RqQ=</me:sig>
  </me:env>
</diaspora>
```

which gets read by the salmon parser (`Salmon::Slap.from_xml`) into this:

```xml
<XML>
  <post>
    <participation>
      <guid>115016cb4edb2eee</guid>
      <target_type>Post</target_type>
      <parent_guid>1e7cab5f6fdd102e</parent_guid>
      <parent_author_signature>ovBw0+xvrzDCJan0maqIbXRMOYuzkrAcAbE4GpW+CCf+eqziTKB8H6lOg6xNAGv7muk1bKxLb5Z1MzCVxvU2z99VJzfnIjZTTthTYwfwjwcLR2BehlrbSZ/k0BDOUI/R2FEMeKXZW6wPYXQFAXaScHqLBFRt0stMPnN0bLBijnMiEQ3Z1gXNaT8TC2Zwf+muEhKvwjD5OvD18iBer/crs/RUsdSREB8FNIqGL1rEx2w4X6QEOmil8XyHf8UZWBPdqYmpG9lAc4ScfMm4iG8/X1owGEOxkRc8MtPPfIryaEvRozhFykinZQZdMi/FXo5ubeQgKOJictX2KBf0YCsDqHChXivLzOGMWRLMQz9EIlds377ZsD19GZKIirE8WpJvFyy/aRWSVBnOSUbUIVvHF+GH24GUGr4nB5rwlkk7pAefL2HE6+/cEafEofdUkgJCtE8fsplQQJC/7RZfVPxMVdI+jrGd8F4Y9Ed0vUu1IRDAJaliIz1NjAT4YGxq70Uxlmp4a6sdAORFd/vD76BN5vCCaeJdHQwiL/VuktXMPR1DFKOJrUL7Sl6EAeuXZhqkLHu9UqYH6METq2+vsFXBtlVMJ2ssumAtiUBADoXxaxLbsxqAbmSbMl/ZbUwJmfNprj1GNWU3vluaHHvmwYNKQIuJIXQUMFjtEGJ/mdj1pYY=</parent_author_signature>
      <author_signature>PNC9jWa4RLNcRYyMZRNRIbcHECxPUmwAcmob/ffZffGzCt4NL7rqNT0NU9rRYrQ6CwUOZ6nmrbDO6HsBVjtyfrtIqrXHod+5cSYRhdOVBWSZ+46bR0SSCAKTbstZUBA4zeXy1WOmMZwCTBwvJe6af86WIIvDOahwWG0uiHCrEnUaNfyZgSi4mMISPOOEDrpC85Kj8L6m1BWDJFfOf3PW52Qkyxl3Xbt8T9UMk3065BagLDru5LGCSFZGRm6z5PcWHflkYxUL7n6zinitxEoR8jMvB5xplxCgIpNsR3AGvTcNBAagqcfsYMtB/+FHHHQ4PJDAg/WodZt0dBRovr9NZe4x/6xQL3E7/g3WejBy+Ay5FsIcnYfyf7fn5tPxnccjX0zbPcF2+UOB+F2uEx9R3DaHizBHTZaECRz0NHj37Nak7X7LOcGWhizhgFzl2Q+kooRulD/zS9KvsAgQk9++0F0GXNwv2CqQZF+YpyxW4UL65PZ1HHd3N01OGdeEEhCcqXyF3qEMLTZB1fBUNow3pul/vONUpawX1+NMRsNBEsEVqNPUdoq/KCx5gSFLw+H5iRqqTCe8vCOQz4al8AnR7U9wEdW6vYM/oK6jHXfybgWOJLzHYPbA0ZvbkzceFWiPYE/RaMf726+yUspr817YUioGW1haufEUOiS/vNClwIE=</author_signature>
      <diaspora_handle>copaflight8@joindiaspora.com</diaspora_handle>
    </participation>
  </post>
</XML>
```

`Diaspora::Parser::from_xml` first creates a Nokogiri document of this xml with 
`doc = Nokogiri::XML(xml) { |cfg| cfg.noblanks }`, and extracts the body with 
`body = doc.xpath("/XML/post").children.first` ... still all is well. The class name is extracted with 
`class_name = body.name.gsub('-', '/')` which is "participation" in our case. 

Now the broken part, we try to create an Instance with: 
`class_name.camelize.constantize.from_xml(body.to_s)`
`class_name.camelize.constantize` is still ok, but 
`from_xml` results in an incomplete object instance:

```
#<Participation 
  id: nil, 
  guid: "115016cb4edb2eee", 
  target_id: nil, 
  target_type: nil, 
  author_id: 5373, 
  author_signature: "PNC9jWa4RLNcRYyMZRNRIbcHECxPUmwAcmob/ffZffGzCt4NL7r...", 
  parent_author_signature: "ovBw0+xvrzDCJan0maqIbXRMOYuzkrAcAbE4GpW+CCf+eqziTKB...", 
  created_at: nil, 
  updated_at: nil>
```

(I don't think `target_id` and `target_type` should be `nil`...)

This in turn makes the recently introduced author-check fail every time, since the `xml_author` method of `Postzord::Receiver::Public` looks like this: 
(`@object` is the Participation instance,
`@object.respond_to? :relayable?` is true and
`@object.parent` is nil)

```ruby
  def xml_author
    if @object.respond_to?(:relayable?)
      #this is public, so it would only be owners sending us other people comments etc
       @object.parent.author.local? ? @object.diaspora_handle : @object.parent.diaspora_handle
    else
      @object.diaspora_handle
    end
  end
```

therefore this exception is raised and the resque job blows up:

```
     undefined method `author' for nil:NilClass

    /srv/diaspora/lib/postzord/receiver/public.rb:67:in `xml_author'
    /srv/diaspora/lib/postzord/receiver.rb:15:in `author_does_not_match_xml_author?'
    /srv/diaspora/lib/postzord/receiver/public.rb:55:in `save_object'
    /srv/diaspora/lib/postzord/receiver/public.rb:26:in `receive!'
    /srv/diaspora/lib/postzord/receiver.rb:11:in `perform!'
    /srv/diaspora/app/models/jobs/receive_unencrypted_salmon.rb:14:in `perform'
```

**Comment by Flaburgan** [comment]
@Raven24 what's the situation of this now that https://github.com/diaspora/diaspora/pull/5209 is merged?

**Comment by denschub** [comment]
Closed by #6873.
